etc/library.sh

```diff
@@ -114,51 +119,151 @@ function run_app_unsafe()
 
   echo "Running $ncp_app"
-  echo "[ $ncp_app ]" >> $log
-
-  # read script
-  unset configure
-  source "$script"
 
-  # read cfg parameters
-  [[ -f "$cfg_file" ]] && {
-    local cfg="$( cat "$cfg_file" )"
-    local len="$(jq '.params | length' <<<"$cfg")"
-    for (( i = 0 ; i < len ; i++ )); do
-      local var="$(jq -r ".params[$i].id"    <<<"$cfg")"
-      local val="$(jq -r ".params[$i].value" <<<"$cfg")"
-      eval "$var=$val"
-    done
+  # Check if app is already running in tmux
+  running_app=$( [[ -f "$LOCK_FILE" ]] && cat "$LOCK_FILE" || echo "" )
+  [[ ! -z $running_app ]] && which tmux > /dev/null && tmux has-session -t="$running_app" > /dev/null 2>&1 && {
+    #echo "Already running. Attaching to output..." | tee -a $log
+    
+    local choice
+    [[ $ATTACH_TO_RUNNING == "1" ]] && choice="y"
+    [[ $ATTACH_TO_RUNNING == "0" ]] && choice="n"
+    question="An app ($running_app) is already running. Do you want to attach to it's output? <y/n>"
+    if [[ $choice == "y" ]] || [[ $choice == "n" ]]
+    then
+      echo "$question"
+      echo "Choice: <y>"
+    else
+      read -rp "$question" choice
+      while [[ "$choice" != "y" ]] && [[ "$choice" != "n" ]]
+      do
+        echo "choice was '$choice'"
+        read -rp "Invalid choice. y or n expected." choice
+      done
+    fi
+
+    if [[ "$choice" == "y" ]]
+    then
+      attach_to_app "$running_app"
+    fi
+    return $?
   }
 
-  # run
-  configure 2>&1 | tee -a $log
-  local ret="${PIPESTATUS[0]}"
-
+  unset configure
+  (
+    # read cfg parameters
+    [[ -f "$cfg_file" ]] && {
+      local len cfg
+      cfg="$( cat "$cfg_file" )"
+      jq -e '.tmux' <<<"$cfg" > /dev/null 2>&1
+      use_tmux="$?"
+      len="$(jq '.params | length' <<<"$cfg")"
+      for (( i = 0 ; i < len ; i++ )); do
+        local var val
+        var="$(jq -r ".params[$i].id"    <<<"$cfg")"
+        val="$(jq -r ".params[$i].value" <<<"$cfg")"
+        eval "export $var=$val"
+      done
+    }
+ 
+    echo "$ncp_app" > "$LOCK_FILE"
+    if which tmux > /dev/null && [[ $use_tmux == 0 ]]
+    then
+      echo "Running $ncp_app in tmux..." | tee -a $log
+      # Run app in tmux
+      local tmux_log_file tmux_status_file LIBPATH
+      tmux_log_file="/var/log/ncp/tmux.${ncp_app}.log"
+      tmux_status_file="/var/log/ncp/tmux.${ncp_app}.status"
+      LIBPATH="$(dirname $CFGDIR)/library.sh"
+      
+      # Reset tmux output
+      echo "[ $ncp_app ]" | tee -a $log
+      echo "[ $ncp_app ]" > "$tmux_log_file"
+      echo "" > "$tmux_status_file"
+      chmod 640           "$tmux_log_file" "$tmux_status_file"
+      chown root:www-data "$tmux_log_file" "$tmux_status_file"
+
+      tmux new-session -d -s "$ncp_app" "bash -c '(
+        trap \"echo \\\$? > $tmux_status_file && rm $LOCK_FILE\" 1 2 3 4 6 9 11 15 19 29
+        source \"$LIBPATH\"
+        source \"$script\"
+        configure 2>&1 | tee -a $log
+        echo \"\${PIPESTATUS[0]}\" > $tmux_status_file
+        rm $LOCK_FILE
```

Comment:
maybe you could add `EXIT` to the trap and save this, not a big deal

Reply:
I thought about this. In the trap I write `$?` to the status file, not the pipe status - do you propose to do it conditionally based on the exit code?